### PR TITLE
[CxxInterop] Add tests for static variables

### PR DESCRIPTION
**Pull request by hlopko**
[SR-12464](https://bugs.swift.org/browse/SR-12464)

### CODE REVIEW

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

File: test/Interop/Cxx/static/Inputs/static-var.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+int makeStaticConst();
+static const int staticConstInit = makeStaticConst();
+
+constexpr int makeStaticConstexpr() { return 32; }
+static constexpr int staticConstexpr = makeStaticConstexpr();
+
+class NonTrivial {
+public:
+  // TODO: Call `NonTrivial(int)` constructor once apple/swift/pull/30630 is
+  // merged.
```
Comment (gribozavr) [inline test/Interop/Cxx/static/Inputs/static-var.h:25]: That PR is only enabling Swift code to call such constructors.

In fact, code below already calls `NonTrivial(1024)` -- so I don't understand what this TODO suggests to do.

File: test/Interop/Cxx/static/Inputs/static-var.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  inline explicit NonTrivial() : val(-1) {}
+  inline explicit NonTrivial(int val) : val(val) {}
+  inline constexpr NonTrivial(int val, int val2) : val(val + val2) {}
+  int val;
+};
+
+static NonTrivial staticInst = NonTrivial(1024);
```
Comment (gribozavr) [inline test/Interop/Cxx/static/Inputs/static-var.h:32]: I'm not sure what `Inst` stands for.

File: test/Interop/Cxx/static/Inputs/static-var.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+constexpr int makeStaticConstexpr() { return 32; }
+static constexpr int staticConstexpr = makeStaticConstexpr();
+
+class NonTrivial {
+public:
+  // TODO: Call `NonTrivial(int)` constructor once apple/swift/pull/30630 is
+  // merged.
+  inline explicit NonTrivial() : val(-1) {}
+  inline explicit NonTrivial(int val) : val(val) {}
+  inline constexpr NonTrivial(int val, int val2) : val(val + val2) {}
```
Comment (gribozavr) [inline test/Interop/Cxx/static/Inputs/static-var.h:28]:
```suggestion
  explicit NonTrivial() : val(-1) {}
  explicit NonTrivial(int val) : val(val) {}
  constexpr NonTrivial(int val, int val2) : val(val + val2) {}
```

File: test/Interop/Cxx/static/static-var-irgen.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+// CHECK: define {{(protected |dllexport )?}}swiftcc i32 @"$s4main13readStaticVars5Int32VyF"()
+// CHECK: [[VALUE:%.*]] = load i32, i32* getelementptr inbounds (%Ts5Int32V, %Ts5Int32V* bitcast (i32* @{{_ZL9staticVar|staticVar}} to %Ts5Int32V*), i32 0, i32 0), align 4
+// CHECK: ret i32 [[VALUE]]
+
+public func writeStaticVar(_ v: CInt) {
+  staticVar = v
+}
+
+// CHECK: define {{(protected |dllexport )?}}swiftcc void @"$s4main14writeStaticVaryys5Int32VF"(i32 %0)
+// CHECK: store i32 %0, i32* getelementptr inbounds (%Ts5Int32V, %Ts5Int32V* bitcast (i32* @{{_ZL9staticVar|staticVar}} to %Ts5Int32V*), i32 0, i32 0), align 4
```
Comment (gribozavr) [inline test/Interop/Cxx/static/static-var-irgen.swift:67]: Please also add tests for passing the `staticVar` inout.

File: test/Interop/Cxx/static/Inputs/static-var.h (hunk 1 above)
Comment (hlopko) [reply test/Interop/Cxx/static/Inputs/static-var.h:25]: Clarified the todo.

File: test/Interop/Cxx/static/Inputs/static-var.h (hunk 2 above)
Comment (hlopko) [reply test/Interop/Cxx/static/Inputs/static-var.h:32]: Renamed.

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

**Comment by gribozavr** [comment]
@swift-ci Please smoke test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

File: test/Interop/Cxx/static/static-var-irgen.swift (hunk 4 above)
Comment (hlopko) [reply test/Interop/Cxx/static/static-var-irgen.swift:67]: Done.

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Review by gribozavr — APPROVED** [review]

**Comment by atrick** [comment]
@hlopko I'm not sure how this failure got through PR testing. Have you already fixed it or do I need to XFAIL the test?

https://ci.swift.org/job/oss-swift_tools-RA_stdlib-DA_test-device-non_executable/70/consoleFull

TEST 'Swift(iphoneos-armv7) :: Interop/Cxx/static/static-var-irgen.swift' FAILED

test/Interop/Cxx/static/static-var-irgen.swift:49:11: error: CHECK: expected string not found in input
 // CHECK: call {{void @_ZN10NonTrivialC[12]Ei\(%class.NonTrivial\* @_ZL16staticNonTrivial, i32 1024\)|%class.NonTrivial\* @"\?\?0NonTrivial@@QEAA@H@Z"\(%class.NonTrivial\* @staticNonTrivial, i32 1024\)}}
           ^
 <stdin>:67:49: note: scanning from here
 define internal void @__cxx_global_var_init.4() #0 section "__TEXT,__StaticInit,regular,pure_instructions" {
                                                 ^

**Comment by hlopko** [comment]
Fix coming in https://github.com/apple/swift/pull/31530. Thanks for reporting.

As for why it was not caught by presubmits, I though we only have macos presubmits, no ios? Maybe @shahmishal can clarify?

**Pull request merged**